src/PieChart/components/PieChartContainer.ts

```diff
@@ -0,0 +1,205 @@
+            errorMessage += ` 'Data source' is set to 'Microflow' but 'Microflow' is missing \n`;
+        }
+
+        return errorMessage && `Configuration error in ${props.chartType}chart:\n\n ${errorMessage}`;
+    }
+
+    private resetSubscriptions(mxObject?: mendix.lib.MxObject) {
+        this.subscriptionHandles.forEach(mx.data.unsubscribe);
+        this.subscriptionHandles = [];
+
+        if (mxObject) {
+            this.subscriptionHandles.push(mx.data.subscribe({
+                callback: () => this.fetchData(mxObject),
+                guid: mxObject.getGuid()
+            }));
+        }
+    }
+
+    private fetchData(mxObject?: mendix.lib.MxObject) {
+        this.colors = []; this.labels = []; this.values = [];
+        if (mxObject && this.props.dataEntity) {
+            if (this.props.dataSourceType === "xpath") {
+                this.fetchByXPath(mxObject);
+            } else if (this.props.dataSourceType === "microflow" && this.props.dataSourceMicroflow) {
+                this.fetchByMicroflow(mxObject.getGuid());
+            }
+        }
+    }
+
+    private fetchByXPath(mxObject: mendix.lib.MxObject) {
+        const constraint = this.props.entityConstraint
+            ? this.props.entityConstraint.replace("[%CurrentObject%]", mxObject.getGuid())
+            : "";
+        const xpath = "//" + this.props.dataEntity + constraint;
+        window.mx.data.get({
+            callback: mxObjects => this.processData(mxObjects),
+            error: error => {
+                mx.ui.error(`An error occurred while retrieving data via XPath (${xpath}): ${error}`);
+                this.setState({ values: [], labels: [], colors: [] });
+            },
+            filter: {
+                sort: [ [ this.props.sortAttribute, "asc" ] ]
+            },
+            xpath
+        });
+    }
+
+    private fetchByMicroflow(guid: string) {
+        const actionname = this.props.dataSourceMicroflow;
+        mx.ui.action(actionname, {
+            callback: mxObjects => this.processData(mxObjects as mendix.lib.MxObject[]),
+            error: error => {
+                mx.ui.error(`Error while retrieving microflow data ${actionname}: ${error.message}`);
+                this.setState({ values: [], labels: [], colors: [] });
+            },
+            params: {
+                applyto: "selection",
+                guids: [ guid ]
+            }
+        });
+    }
+
+    private processData(mxObjects: mendix.lib.MxObject[]) {
+        mxObjects.map(value => {
+            this.values.push(parseFloat(value.get(this.props.valueAttribute) as string));
+            this.labels.push(value.get(this.props.nameAttribute) as string);
+            this.colors.push(value.get(this.props.colorAttribute) as string);
+        });
+        this.setState({
+            colors: this.colors,
+            labels: this.labels,
+            values: this.values
+        });
```

Comment:
As far as I can tell, these properties (`colors, labels, values`) are only useful in this function. You can have the same functionality with them created locally in this file. Come we do it when u get to this point.... biwanvu at the moment.